# Defining your own functions

You have seen builtin functions, like type(), int(), str(), and functions from packages, like math.sqrt(). In Python, you can also define functions yourself. This is useful when you need to do the same thing (or something very similar) again and again. Suppose that you are in a country where temperature is given in Celsius, but you are more familiar with Fahrenheit. Here is how you can figure out that 20 degrees Celsius are 68 degrees Fahrenheit, and 30 degrees Celsius are 86 degrees Fahrenheit:

In [1]:
print( (20 * 9/5) + 32)

68.0


In [2]:
print( (30 * 9/5) + 32)

86.0


If you need to convert temperatures very often, this gets tedious. Instead, you can define your own function by giving a name to a piece of code:

In [3]:
def celsius_2_fahrenheit(temp):
    fahrenheit = temp * 9/5 + 32
    return fahrenheit

In [4]:
# here is a variant without the variable "fahrenheit"
def celsius_2_fahrenheit_variant(temp):
    return (temp * 9/5) + 32

After that, you can use ```celsius_2_fahrenheit``` like any built-in function:

In [5]:
celsius_2_fahrenheit(38)

100.4

In [6]:
celsius_2_fahrenheit(0)

32.0

In [7]:
celsius_2_fahrenheit(100)

212.0

In [8]:
fahrenheit_val = celsius_2_fahrenheit(25)
type(fahrenheit_val)

float

In [9]:
print(fahrenheit_val)

77.0


In [10]:
# if you call your function with a parameter value
# that doesn't match what you do with the parameter
# in the function, you get an error, as usual

# celsius_2_fahrenheit("ten")

In [11]:
# "ten" * 9

In [12]:
# 'tententententententententen' / 5

# Input and output

A function has an input and an output. You know that from a builtin function like ```len()```: It takes as input an object, for example a list. You put the input between the parentheses. And it returns an output, the length of the object. You can, for example, store that output in a variable:

In [13]:
len("how are you doing")

17

In [14]:
mylist = ["a", "b", "c"]
somevariable = len(mylist)
print("the output of len that I stored is", somevariable)

the output of len that I stored is 3


The same is true for functions that you define yourself. You define the inputs in parentheses after the function name. The function name is ```celsius_2_fahrenheit```, and in parentheses after that you see ```(temp)```. What is ```temp```? It is just a variable name. When you call a pre-defined function, you give it an actual input value. When you define your own function, you just prepare a container, a variable name, that will store the input. In this case, that container is ```temp```.

The output of the function is what you see after the keyword ```return```: It is the value currently stored in ```fahrenheit```.

So when we call our self-defined function with a different input, for example 100 (that is, 100 degrees celsius), the output that we get is 212:

In [15]:
celsius_2_fahrenheit(100)

212.0

The variable ```temp``` in the function definition is used in much the same way as in function definitions in mathematics:

f(x) = x+1

There, you don't have to specify beforehand what x is. Rather, for each x, the function value is f(x) = x+1. In the same way, when you call your function saying celsius_2_fahrenheit(20), then at this time, Python stores the value 20 in the variable ```temp``` and executes the code of the function with ```temp``` set to 20.


# The components of a function definition

Let's take a look at all the bits and pieces of this function definition:

In [16]:
def celsius_2_fahrenheit(temp):
    fahrenheit = temp * 9/5 + 32
    return fahrenheit

* ```def``` is a reserved word. It tells Python that you are defining a function.

* The name of the function is ```celsius_2_fahrenheit```. This is a variable that you define, just this time it is a container that contains Python code, not a number or a string. The names of functions are subject to the same restrictions as other Python variable names.

* After the name of the function, you see the input to the function, also called its argument, in round brackets. This is a variable name: the container that we have prepared to take the actual input when the function is called.

* After the argument there is a colon. 
* Then come some indented lines, which define what the function does.

In [17]:
# compare to:
for i in ["a", "b" "c"]:
    print(i)
    

a
bc


So we have an overall shape of 

```keyword something:
    indented code```
    
which is the same as, for example, in for-loops and in if-conditions.

**Try it for yourself:**

* Define a function that converts degrees Fahrenheit to degrees Celsius (for the  benefit of Europeans like me). 

* Define a function that takes a string as input, then strips all punctuation symbols from the beginning and end and lowercases. For example, when the input is "Hello!" the output should be "hello", and when the input is "???WHAT???", the output should be "what". 

In [18]:
from string import punctuation as punc

def fahrenheit_to_celsius(temp):
    return((temp - 32) * 5/9)

def strip_string(string):
    return string.strip(punc).lower()

print(fahrenheit_to_celsius(50))
print(strip_string("???WHAT???"))

10.0
what


# more about "return"

"return" means: hand back this value, and you're done. Any line after this one in the function will not be executed anymore:

In [19]:
def c2f(temp):
    fahrenheit = (temp * 9/5) + 32
    print("about to return value")
    return fahrenheit
    print("after returning value")

c2f(20)

about to return value


68.0

But you can, of course, use "return" in an "if" statement:

In [20]:
def maybe_lowercase(word):
    if word.startswith("A"):
        return word.lower()
    else:
        return word

print(maybe_lowercase("AARDVARK"))
print(maybe_lowercase("ZEBRA"))

aardvark
ZEBRA


## Returning lists

The output, what you write after the keyword ```return```, can be a single value: a number, as in the Celsius-to-Fahrenheit example, or a string, as in the stripping-punctuation-and-lowercasing example. The output of a function can also be a list. Here is a function that takes a string as input, tokenizes it, lowercases, and removes punctuation:

In [21]:
import string
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [22]:
"hello!!!".strip(string.punctuation)

'hello'

In [23]:
w = "!!?!!"
w.strip(string.punctuation) == ""

True

In [24]:
import nltk
import string

def my_preprocess(inputstring):
    words = nltk.word_tokenize(inputstring)
    newwords = [w.lower() for w in words if w.strip(string.punctuation) != ""]
    return newwords

my_preprocess("What a lovely day!!")

['what', 'a', 'lovely', 'day']

When you want to return multiple pieces of data, you can also return them in a list, or in a tuple.

You can then catch the multiple return values in multiple separate variables:

In [25]:

def my_preprocess(inputstring):
    words = nltk.word_tokenize(inputstring)
    newwords = [w.lower() for w in words if w.strip(string.punctuation) != ""]
    punctuations = [w for w in words if w.strip(string.punctuation) == ""]
    # this makes a tuple of words without punctuation, punctuation symbols
    return (newwords, punctuations)

words, punctuationsymbols = my_preprocess("What a lovely day!!")
print(words)
print(punctuationsymbols)

['what', 'a', 'lovely', 'day']
['!', '!']


In [26]:
my_preprocess("What a lovely day!!")


(['what', 'a', 'lovely', 'day'], ['!', '!'])

**Try it for yourself:**

* Define a function that takes as input a string, splits it on whitespace, removes every word that starts with "http", and returns the remaining list of words. So if the input is ```"Go to https:\\abc.def.gh"```, the output should be ```["Go", "to"]```

* Define a function that takes as input a string of numbers separated by whitespace, for example ```"123 3.4  67.9"```. It should return a list with the numbers that were in the string, in our case ```[123, 3.4, 67.9]```.`

In [27]:
import re

def str_func(input):
    lst1 = input.split()
    
    for word in lst1:
        if re.search(r"^https", word):
            lst1.remove(word)
    
    return lst1

def int_func(input):
    lst1 = input.split()

    lst2 = []

    for item in lst1:
        lst2.append(float(item))

    return lst2


int_func("123 3.4  67.9")

[123.0, 3.4, 67.9]

# Functions with multiple arguments

A function can have more than one piece of input. Here is an example. This function repeats a string a given number of times:

In [28]:
"x" * 10

'xxxxxxxxxx'

In [29]:
"x" + "y"

'xy'

In [30]:
def repeat_string(somestring, numtimes):
    return somestring * numtimes

repeat_string("hello", 5)

'hellohellohellohellohello'

In [31]:
repeat_string("hello", 2)

'hellohello'

# Optional arguments

You can define a function with optional arguments. For an argument that is optional, you give a *default* value in the definition. When calling the function, you can either keep or overwrite the default.

In [32]:
def repeat_string_again(somestring, numtimes = 5):
    return somestring * numtimes

In [33]:
repeat_string_again("hello", 10)

'hellohellohellohellohellohellohellohellohellohello'

In [34]:
repeat_string_again("hello")

'hellohellohellohellohello'

**Important: When you use optional arguments, you have to put all the non-optional arguments before the optional ones in your function definition.**`

In [35]:
# this will not work
# error: SyntaxError: non-default argument follows default argument

# def repeat_string_withbug(numtimes = 5, somestring):
    # return somestring * numtimes

# Put all values that the function uses in the list of input parameters

Make your function so that its list of input parameters is its only "service window", the only way to pass in values. 

In [36]:
# in principle you can do this

numtimes_baddesign  = 5

def repeatstring(s):
    return s * numtimes_baddesign


In [37]:
# but please do this

def repeatstring_better(s, numtimes):
    return s * numtimes

# Using functions when transforming Pandas columns

In the Python list comprehensions notebook, I said that list comprehensions are great for transforming a Pandas dataframe column. Here is the example from that notebook again: 

In [38]:
import pandas as pd

mydf = pd.DataFrame( {"line" : ["How doth the little crocodile",
                                "Improve his shining tail", 
                                "And pour the waters of the Nile",
                                "On every golden scale!",
                                "How cheerfully he seems to grin",
                                "How neatly spreads his claws,",
                                "And welcomes little fishes in",
                                "With gently smiling jaws"]})

mydf

line
0    How doth the little crocodile
1         Improve his shining tail
2  And pour the waters of the Nile
3           On every golden scale!
4  How cheerfully he seems to grin
5    How neatly spreads his claws,
6    And welcomes little fishes in
7         With gently smiling jaws

In [39]:
# Then we added a new column with 
# tokenized lines like this, using
# a list comprehension:
mydf["tokenized"] = [nltk.word_tokenize(s) for s in mydf.line]
mydf

line                                tokenized
0    How doth the little crocodile      [How, doth, the, little, crocodile]
1         Improve his shining tail            [Improve, his, shining, tail]
2  And pour the waters of the Nile  [And, pour, the, waters, of, the, Nile]
3           On every golden scale!            [On, every, golden, scale, !]
4  How cheerfully he seems to grin   [How, cheerfully, he, seems, to, grin]
5    How neatly spreads his claws,    [How, neatly, spreads, his, claws, ,]
6    And welcomes little fishes in      [And, welcomes, little, fishes, in]
7         With gently smiling jaws            [With, gently, smiling, jaws]

But not all column transformations are this simple. Say we want to do more preprocessing: Tokenize the line, lowercase, and remove stopwords.

This is clunky when expressed in a single list comprehension. But we can define a function that does such a transformation, then use that in a list comprehension:

In [40]:
import nltk
stopwords = nltk.corpus.stopwords.words("english")

def my_preprocess(line):
    words = nltk.word_tokenize(line)
    words1 = [w.lower() for w in words]
    words2 = [w for w in words1 if w not in stopwords]
    return words2


mydf["preprocessed"] = [my_preprocess(line) for line in mydf.line]
mydf

line                                tokenized  \
0    How doth the little crocodile      [How, doth, the, little, crocodile]   
1         Improve his shining tail            [Improve, his, shining, tail]   
2  And pour the waters of the Nile  [And, pour, the, waters, of, the, Nile]   
3           On every golden scale!            [On, every, golden, scale, !]   
4  How cheerfully he seems to grin   [How, cheerfully, he, seems, to, grin]   
5    How neatly spreads his claws,    [How, neatly, spreads, his, claws, ,]   
6    And welcomes little fishes in      [And, welcomes, little, fishes, in]   
7         With gently smiling jaws            [With, gently, smiling, jaws]   

                  preprocessed  
0    [doth, little, crocodile]  
1     [improve, shining, tail]  
2         [pour, waters, nile]  
3    [every, golden, scale, !]  
4    [cheerfully, seems, grin]  
5  [neatly, spreads, claws, ,]  
6   [welcomes, little, fishes]  
7      [gently, smiling, jaws]

**Try it for yourself**: 

* Can you change the function ```my_preprocess()``` such that it also filters away words that consist solely of punctuation?


In [41]:
import nltk
from string import punctuation as punc

stopwords = nltk.corpus.stopwords.words("english")

import pandas as pd

mydf = pd.DataFrame( {"line" : ["How doth the little crocodile",
                                "Improve his shining tail", 
                                "And pour the waters of the Nile",
                                "On every golden scale!",
                                "How cheerfully he seems to grin",
                                "How neatly spreads his claws,",
                                "And welcomes little fishes in",
                                "With gently smiling jaws"]})

mydf

def my_preprocess(line):
    words = nltk.word_tokenize(line)
    words1 = [w.lower() for w in words]
    words2 = [w for w in words1 if w not in stopwords]
    words3 = [w for w in words2 if w.strip(punc) != ""]
    return words3

mydf["tokenized"] = [nltk.word_tokenize(s) for s in mydf.line]
mydf["preprocessed"] = [my_preprocess(line) for line in mydf.line]
mydf

line                                tokenized  \
0    How doth the little crocodile      [How, doth, the, little, crocodile]   
1         Improve his shining tail            [Improve, his, shining, tail]   
2  And pour the waters of the Nile  [And, pour, the, waters, of, the, Nile]   
3           On every golden scale!            [On, every, golden, scale, !]   
4  How cheerfully he seems to grin   [How, cheerfully, he, seems, to, grin]   
5    How neatly spreads his claws,    [How, neatly, spreads, his, claws, ,]   
6    And welcomes little fishes in      [And, welcomes, little, fishes, in]   
7         With gently smiling jaws            [With, gently, smiling, jaws]   

                 preprocessed  
0   [doth, little, crocodile]  
1    [improve, shining, tail]  
2        [pour, waters, nile]  
3      [every, golden, scale]  
4   [cheerfully, seems, grin]  
5    [neatly, spreads, claws]  
6  [welcomes, little, fishes]  
7     [gently, smiling, jaws]